File: chap07_calcium/RyR_Stern.m

%  -------------------------------------------------------------------
%
%   Reponse of the RyR model of Stern to a step increase in calcium.
%
%   For Chapter 7, Section 7.4.4 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
% 
%   Written by James Keener and James Sneyd.
% 
%  ------------------------------------------------------------------- 


%% symbolic calculation and plotting of the steady state

clear all
close all
clc
set(0,                           ...
   'defaultaxesfontsize', 18,   ...
   'defaultaxeslinewidth', 2.0, ...
   'defaultlinelinewidth', 2.0); 

syms k1 km1 k2 km2 R O I RI c
global k1val km1val k2val km2val

RI = 1 - R - O - I;
eq1 = km1*O + km2*RI - R*(k1*c*c + k2*c);
eq2 = km2*I + k1*c*c*R - O*(k2*c + km1);
eq3 = k2*c*O + k1*c*c*RI - I*(km2 + km1);

k1val = 35;
km1val = 0.06;
k2val = 0.5;
km2val = 0.01;

[R,O,I] = solve([eq1,eq2,eq3],[R,O,I]);
Ost(c) = subs(O,[k1,km1,k2,km2],[k1val,km1val,k2val,km2val]);
Rst(c) = subs(R,[k1,km1,k2,km2],[k1val,km1val,k2val,km2val]);
Ist(c) = subs(I,[k1,km1,k2,km2],[k1val,km1val,k2val,km2val]);

figure(2)
fplot(Rst,[0.01 0.5],'r','LineWidth',2)
hold on
fplot(Ost,[0.01 0.5],'b','LineWidth',2)
fplot(Ist,[0.01 0.5],'g','LineWidth',2)

% for external plotting
% keep = fplot(Ost,[0.01 0.5]);
% writematrix([keep.XData' keep.YData'],'test1.dat')

%% numerical solutions

%clear all

init = [1,0,0];                 % Assume that c = 0 initially. Probably not very accurate.
tspan = linspace(0,1.5,1000);

figure(1)
c = 1;
[T,Y] = ode15s(@(t,x)rhs(t,x,c),tspan,init);
plot(T,Y(:,2))
%writematrix( [T Y(:,2)],'test3.dat')   % for external plotting
hold on

c = 2;
[T,Y] = ode15s(@(t,x)rhs(t,x,c),tspan,init);
%writematrix( [T Y(:,2)],'test4.dat')   % for external plotting
plot(T,Y(:,2))

c = 3;
[T,Y] = ode15s(@(t,x)rhs(t,x,c),tspan,init);
%writematrix( [T Y(:,2)],'test5.dat')   % for external plotting
plot(T,Y(:,2))
hold off

% calculate the steady-state and peak responses

n = 50;
cc = linspace(0.01,0.5,n);
tspan = linspace(0,50,100000);
for i = 1:n
    c = cc(i);
    [T,Y] = ode15s(@(t,x)rhs(t,x,c),tspan,init);
    peak(i) = max(Y(:,2));
end

figure(2)
plot(cc,peak,'k')
 
% for external plotting
% writematrix([cc',peak'],'test2.dat')



%%
function out = rhs(t,x,c)
global k1val km1val k2val km2val
R = x(1);
O = x(2);
I = x(3);
RI = 1 - R - O - I;

k1 = k1val;
km1 = km1val;
k2 = k2val;
km2 = km2val;

out(1) = km1*O + km2*RI - R*(k1*c*c + k2*c);
out(2) = km2*I + k1*c*c*R - O*(k2*c + km1);
out(3) = k2*c*O + k1*c*c*RI - I*(km2 + km1);
out = out';

end

File: chap07_calcium/RyR_Stern.py
#   -------------------------------------------------------------------
# 
#    Reponse of the RyR model of Stern to a step increase in calcium.
# 
#    For Chapter 7, Section 7.4.4 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd.
# 
#   -------------------------------------------------------------------


import sympy as sp
import matplotlib.pyplot as plt
import numpy as np
from scipy.integrate import solve_ivp


# Calculate the steady states for any given c
def ROIss(c):
    A = np.array([[-km2 - (k1 * c**2 + k2 * c), km1 - km2, -km2],
                  [k1 * c**2, -(k2 * c + km1), km2],
                  [-k1 * c**2, -k1 * c**2 + k2 * c, -(km2 + km1) - k1 * c**2]])
    
    Rhs = np.array([-km2, 0, -k1 * c**2])
    R = np.linalg.solve(A, Rhs)
    return R[0], R[1], R[2]


# Define the ODEs
def rhs(t, x, c):
    R, O, I = x
    RI = 1 - R - O - I

    return [
        km1 * O + km2 * RI - R * (k1 * c * c + k2 * c),
        km2 * I + k1 * c * c * R - O * (k2 * c + km1),
        k2 * c * O + k1 * c * c * RI - I * (km2 + km1)
    ]


k1 = 35
km1 = 0.06
k2 = 0.5
km2 = 0.01

# Plot the steady states as functions of c
clist = np.linspace(0.01, 0.5, 50)
Rst, Ost, Ist = [], [], []
for c in clist:
    R, O, I = ROIss(c)
    Rst.append(R)
    Ost.append(O)
    Ist.append(I)

plt.figure(1)
plt.plot(clist, Rst, 'r', label='Rst')
plt.plot(clist, Ost, 'b', label='Ost')
plt.plot(clist, Ist, 'g', label='Ist')
plt.legend()
plt.xlabel('c (\u03BCM)')
plt.ylabel('Steady states')
plt.title('Steady states vs. c')



# Next, do the numerical solutions and calculate the peak response

# Set initial conditions, time span, and parameter values
init = [1, 0, 0]  # Initial conditions for R, O, I
tspan = np.linspace(0, 1.5, 1000)  # Time span
c_vals = [1, 2, 3]  # Values of c to solve for

# Solve the ODEs for different values of c
plt.figure(2)
for c in c_vals:
    sol = solve_ivp(lambda t, x: rhs(t, x, c), (tspan[0], tspan[-1]), init, t_eval=tspan)
    plt.plot(sol.t, sol.y[1], label=f'c = {c}')

plt.xlabel('Time')
plt.ylabel('O')
plt.legend()

# Calculate and plot the peak response
cc = np.linspace(0.01, 0.5, 50)
tspan = np.linspace(0, 50, 100000)
peak = []
for c in cc:
    sol = solve_ivp(lambda t, x: rhs(t, x, c), (tspan[0], tspan[-1]), init, t_eval=tspan)
    peak.append(sol.y[1].max())

plt.figure(1)
plt.plot(cc, peak, 'k',label='O (peak)')
plt.xlabel('c')
plt.legend()


